### PR TITLE
custom interactions for partially fillable orders

### PR DESCRIPTION
**Pull request by MartinquaXD**
As requested by @alfetopito and @anxolin this PR adds rudimentary support for custom interactions for partially fillable limit orders. This is needed to fully support completely gasless `permi`-able token trades. [context](https://cowservices.slack.com/archives/C0361CDD1FZ/p1695649451392219)

So far we didn't support custom interactions for partially fillable orders because their semantics are somewhat ambiguous (e.g. only execute on first fill, when filled completely, on every fill). Ideally there would be a mechanism for the user to tell us when they want their custom interactions to be executed but I think only executing them on the very first fill seems like a good default which can later be replaced by a more sophisticated approach.

### CODE REVIEW

File: crates/driver/src/infra/api/routes/solve/dto/auction.rs [hunk 1]
```diff
@@ -64,24 +64,34 @@ impl Auction {
...
-                        post_interactions: order
-                            .post_interactions
-                            .into_iter()
-                            .map(|interaction| eth::Interaction {
-                                target: interaction.target.into(),
-                                value: interaction.value.into(),
-                                call_data: interaction.call_data.into(),
-                            })
-                            .collect(),
+                        pre_interactions: match order.executed.is_zero() {
```
Comment (nlordell) [inline crates/driver/src/infra/api/routes/solve/dto/auction.rs:67]: Not sure if you think this is nicer or not, but:

```suggestion
                        pre_interactions: order.executed.is_zero()
                            .then(|| order.pre_interactions...)
                            .unwrap_or_default(),
```

**Review by nlordell — COMMENTED** [review]
Code looks good - buy can you update the `openapi.yaml` to remove the error condition (if its even there 😛). Also, would be nice to have an E2E test to verify this.

Another comment is that this is conflating hooks with pre-interactions. For example, if we decide to use a pre-interaction to warm up storage for transferring Ether to a smart contract instead of access lists, then this will break it. Maybe the check should be more explicit: "include **hooks** if and only if the order is completely un-executed".

File: crates/solver/src/auction_preprocessing.rs [hunk 2]
```diff
@@ -1,10 +1,31 @@
...
+/// already been executed. We do this to ensure that each interaction gets only
+/// executed once.
+pub fn filter_executed_custom_interactions(orders: &mut [Order]) {
+    let zero = 0u32.into();
+
+    let was_partially_executed = |order: &Order| {
+        order.metadata.executed_buy_amount != zero
+            || order.metadata.executed_sell_amount != zero
+            || order.metadata.executed_sell_amount_before_fees != U256::zero()
+            || order.metadata.executed_fee_amount != U256::zero()
```
[the remark this answers is not in the document]
Comment (MartinquaXD) [reply crates/solver/src/auction_preprocessing.rs:22]: I believe there is no good reason to have different types here. `U256` should be the correct type for all fields.

File: crates/driver/src/infra/api/routes/solve/dto/auction.rs [hunk 3]
```diff
@@ -64,24 +64,34 @@ impl Auction {
...
+                                    target: interaction.target.into(),
+                                    value: interaction.value.into(),
+                                    call_data: interaction.call_data.into(),
+                                })
+                                .collect(),
+                            // custom interactions of partially fillable orders should only be
+                            // executed on the first fill
+                            false => vec![],
+                        },
+                        post_interactions: match order.executed.is_zero() {
```
Comment (fleupold) [inline crates/driver/src/infra/api/routes/solve/dto/auction.rs:81]: I can see how pre-interactions may only have to be executed on the first fill. However, for post interactions, I'm not sure this is so clear. The use case we envision for post interactions are things like bridging, staking, etc which could happen after every fill and if only once, they should happen on the last fill (which is impossible to predict)

File: crates/solver/src/auction_preprocessing.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (fleupold) [reply crates/solver/src/auction_preprocessing.rs:22]: Note, that the executed buy/sell amounts are not u32 (as one may think from just looking at this code), but BigUints (but true that U256 would be enough and more consistent).

**Review by fleupold — APPROVED** [review]
LGTM. Agree that it would be nice to differentiate between hooks and generic interactions and also would like to discuss if post interactions shouldn't be executed on every fill?

File: crates/driver/src/infra/api/routes/solve/dto/auction.rs (hunk 3 above)
Comment (MartinquaXD) [reply crates/driver/src/infra/api/routes/solve/dto/auction.rs:81]: Hmm, that's a good point. I guess we would only know the last fill for sure when encoding the solution (executed_amount_after_fill == total_amount) and only then make that part of the solution. On the other hand this would mess with the fee computation as we would somehow have to make solvers aware that there are additional costs for filling completely.
Also AFAIK some solvers currently have the problem that they leave dust in partially fillable orders meaning that it's uneconomic for other solvers to fill the remaining dust to actually trigger the post interaction.

Comment (nlordell) [reply crates/driver/src/infra/api/routes/solve/dto/auction.rs:81]: We can change the PR to allow pre-interactions for all order types (but keep post-interactions disabled for partially fillable orders)?

**Review by anxolin — APPROVED** [review]

File: crates/driver/src/infra/api/routes/solve/dto/auction.rs (hunk 3 above)
Comment (MartinquaXD) [reply crates/driver/src/infra/api/routes/solve/dto/auction.rs]: I now went for pre-interactions only on first fill and post-interactions on every fill.
Because there is currently no mechanism to tell an interaction how much of an order will be filled the hook basically has to prepare for the "worst case" from the perspective of the hook.
For pre-interactions that means it has to assume it's a 100% fill and needs to produce all the `sell_token` to support the whole order. It only makes sense to do that once.

For post-interactions the situation is not so clear cut. IMO the only sensible options are (1) execute after last fill or (2) execute after every fill. Because some edge cases might result in an order not being filled 100% I think we should opt for `2`. Additionally from the perspective of a hook implementer writing a hook compatible with `2` is the safest option and since we don't have any on-chain guarantees for the execution of hooks implementers should prepare for the worst to be safe.
AFAICS the only consequence of this is that the hook should be written in a way that it checks the fill status of the order to determine the correct action.
So in conclusion although I initially thought there are multiple viable approaches for executing interactions for partial fills I now believe there is only 1 reasonable decision here which means we don't have to add any sophisticated mechanism to let the user decide how/when their interactions get executed (can't be enforced anyway so why bother).

File: crates/driver/src/infra/api/routes/solve/dto/auction.rs (hunk 1 above)
Comment (MartinquaXD) [reply crates/driver/src/infra/api/routes/solve/dto/auction.rs]: I moved that logic into the `autopilot` and used your suggestion there. 👍

**Comment by MartinquaXD** [comment]
I think the PR is in a good shape but should still get an e2e test. I'll try to finish that today but I'm not optimistic. In case I don't manage to do that and we want to have `permit` work before I get back someone else would have to finish this while I'm gone.

**Comment by fleupold** [comment]
Looks good. You can also leave the e2e test up for me (I want to take some time to understand our e2e testing framework next week and am still looking for a good use case)

**Review by nlordell — APPROVED** [review]

File: crates/driver/src/infra/api/routes/solve/dto/auction.rs (hunk 3 above)
Comment (anxolin) [reply crates/driver/src/infra/api/routes/solve/dto/auction.rs]: Until we have people demanding for it, I wouldn't do it. But would we want to enhance the semantics on how we define pre/post hooks to give an indication on how to execute this? 

This can be just a soft contract, the protocol don't enforce it, but is still nice to have this flexibility.

In principle, post-interactions could be even done in another transaction. So one potential solution to post-hooks for partial fills, is to do it in the next block (or next settlement) once we know the order is 100% filled.  I don't think this is good in every case, but its possibly a construct that is useful for many cases (like TWAP and bridge at the end).

Comment (fleupold) [reply crates/driver/src/infra/api/routes/solve/dto/auction.rs]: >Until we have people demanding for it, I wouldn't do it.

Why? We are on it right now and I follow Martin's logic that there is only one good way to implement it, so why not include it? When someone needs it we need to make ressources for it, switch context and accompany the change in production which will take time.

> So one potential solution to post-hooks for partial fills, is to do it in the next block

I don't think there is any value in delaying post hook execution into the next block (as it means it could potentially frontrun by other actors). By the time we craft the settlement tx, we know as much as we know in the next block so we may as well include it right away. In particular we won't know, if an order is 99% filled, if it will ever get another fill or not (may depend on gas and price swings) so we can never really decide when the last fill happened.

Comment (anxolin) [reply crates/driver/src/infra/api/routes/solve/dto/auction.rs]: > Why? We are on it right now and I follow Martin's logic that there is only one good way to implement it, so why not include it? When someone needs it we need to make ressources for it, switch context and accompany the change in production which will take time.

I think there's a misunderstanding. I can see my message was a bit misleading. I agree with the proposed solution, what I want to say, is that we might want to enhance the `appData` definition. More specifically we can add more information in the Hook definition on how the user wants to execute the pre/post interactions: see https://github.com/cowprotocol/app-data/blob/main/src/schemas/hook/v0.1.0.json

With "Until we have people demanding for it," i was referring to doing this enhancement to allow users to be more specific on how they want their hooks executed. Martin's proposal makes a great default, and I would love to see it implemented

Comment (anxolin) [reply crates/driver/src/infra/api/routes/solve/dto/auction.rs]: > I don't think there is any value in delaying post hook execution into the next block (as it means it could potentially frontrun by other actors). By the time we craft the settlement tx, we know as much as we know in the next block so we may as well include it right away. In particular we won't know, if an order is 99% filled, if it will ever get another fill or not (may depend on gas and price swings) so we can never really decide when the last fill happened.

For me the posthook default makes sense when is 100%. We don't know if it will ever be executed if there's some 1% missing, but we know when the order expires. 
I think is a fair pragmatic default to:
- Execute only when is 100% (if dust is a problem we might want to allow a tiny error tolerance threshold)
- Also execute if the order is expired and partially filled

**Comment by fleupold** [comment]
@alfetopito what's the timeline on this? Martin is off this week. Should we commandeer and merge it or can it wait until he is back next week?

**Comment by alfetopito** [comment]
> @alfetopito what's the timeline on this? Martin is off this week. Should we commandeer and merge it or can it wait until he is back next week?

The sooner we have it the better as it gives us more time to test the flow.
There's a pending PR where this can be tried already https://github.com/cowprotocol/cowswap/pull/3164

There are [other issues pending](https://github.com/cowprotocol/cowswap/issues?q=is%3Aopen+label%3Apermit+is%3Aissue) before we are ready for a release (flagged with `mvp`).
I hope I can have them done at most by this Thursday.

So, given that there are enough approvals, any concerns if this is merged already?

**Comment by sunce86** [comment]
Sorry for late comment, but do we need this for colocated driver also?
